File: Problem1_GroupAnagrams.cpp
// Time Complexity : O(nklogk) k is string length and n is number of strings
// Space Complexity : O(nk) n being number of strings and each having a k number of strings in vector
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        
        if(strs.empty()) return vector<vector<string>>{} ;
        
        unordered_map<string,vector<string>> map;
        
        for(string s : strs){
            // int n=s.length();
            // char charVector[n+1] ; 
            // strcpy(charVector, s.c_str());
            
            string t = s;
            sort(t.begin(),t.end());
            
            if(!map.count(t)){
                map.insert({t,vector<string>{}}); 
            }
//             auto it=map.find(t);
            
//             it->second.insert(
            
            map[t].push_back(s);

        }
        
//         vector<vector<string>> res;
        
//         for (auto it = map.begin(); it != map.end(); it++) {
//         //sort(it->second.begin(), (*it).second.end());
//         res.push_back(it->second);
//         } 
       
//         return res;
        
        vector<vector<string>> anagrams;
        for (auto it : map) { 
            anagrams.push_back(it.second);
        }
        return anagrams;
        
        
    }
};

File: Problem1-GroupAnagrams.java
// Time Complexity : O(nk) where n is the number of words in strs and k is the average length of each word
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        int prime[] = {2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97,101};
        HashMap<Double, List<String>> anaMap = new HashMap<>();
        for(String s: strs)
        {   double res = 1;
            for(int i =0;i< s.length();i++)
            {
                res*= prime[s.charAt(i)-'a'];
            }
         List<String> arr = anaMap.getOrDefault(res, new ArrayList<String>());
         arr.add(s);
         anaMap.put(res,arr);
            
        }
        
        return new ArrayList<>(anaMap.values());
    }
}